### ISSUE TITLE
Clean up SCTP tests

### ISSUE THREAD

**Issue by danwinship** [kubernetes/kubernetes#96717]
When we first added SCTP tests, we thought that we weren't going to be able to do actual tests of SCTP network connectivity within kubernetes CI, so we came up with a plan involving a split between tests that involve actual SCTP network connectivity (which require the SCTP kernel module to be available) and tests that don't do SCTP network traffic and so can run anywhere. That way we could still test some basic functionality even if we couldn't test that SCTP _actually_ worked. Then we got a little bit crazy with the non-connectivity-checking tests, poking around at iptables to try and confirm they were doing the right thing even though we couldn't send SCTP packets to test them the right way.

In fact, it turns out we can run the connectivity tests in kube CI under `kind`, so the split wasn't really necessary and should be cleaned up now.

#### I. Flip the use of `[Disruptive]`

The non-connectivity-checking (`[Feature:SCTP]`) tests conflict with the connectivity-checking (`[Feature:SCTPConnectivity]`) ones, because the former tests use `CheckSCTPModuleLoadedOnNodes` to verify that running the test itself doesn't cause the `sctp.ko` kernel module to be loaded on any nodes, while the latter tests may cause the module to be loaded as a side effect of being the first process on a node to open an SCTP socket.

In the current code, we handle this by marking the `[Feature:SCTPConnectivity]` tests as `[Disruptive]` but now I think this is backward; while it's true that it's the connectivity tests that have a side effect, it's only a "disruptive" side effect because the kernel-module-checking tests are inherently non-parallel-testing-friendly. So we should remove the `[Disruptive]` from the `[Feature:SCTPConnectivity]` tests, and add it to the non-connectivity tests, as described below.

#### II. Simplify the core `[Feature:SCTP]` tests

The three `[Feature:SCTP]` tests in `test/e2e/network/service.go` ("should allow creating a basic SCTP service with pod and endpoints", "should create a ClusterIP Service with SCTP ports", and "should create a Pod with SCTP HostPort") are doing three things:

 1. making sure the API works
 2. poking around at iptables to see if kube made the rules we think it should have made
 3. confirming that running the test doesn't cause the kernel to load the `sctp.ko` kernel module

1 is completely redundant with the "connectivity" tests. 2 is unnecessary given the connectivity tests, and very fragile anyway. 3 is still important (for both functionality and security reasons; see the KEP for details).

The three tests should be replaced with a single simpler test, something like "should allow creating Pods, Services, and NetworkPolicies with SCTP ports". (It probably doesn't belong in `service.go` any more at this point though.) It doesn't try to _use_ the Pod/Service/NetworkPolicy it creates, it just confirms that (a) you can create them, and (b) the result of `CheckSCTPModuleLoadedOnNodes` doesn't change from the start of the test to the end of the test

And while we're fixing things, the new combined test doesn't need to be labelled `[Feature:SCTP]` any more, since SCTP is no longer gated.

#### III. Simplify "should not allow access by TCP when a policy specifies only SCTP" just a little

The one test in `test/e2e/network/network_policy.go` that is `[Feature:SCTP]` rather than `[Feature:SCTPConnectivity]` is "should not allow access by TCP when a policy specifies only SCTP". This makes sure network plugins that _don't_ implement SCTP don't misinterpret policies containing `protocol: SCTP`, so we don't want to move it into the `[Feature:SCTPConnectivity]` suite (since we specifically want this test to be run by some people who are intentionally skipping the `[Feature:SCTPConnectivity]` tag).

So it should stay mostly as is, except that, like above, it doesn't need to be labelled `[Feature:SCTP]` any more. And we can remove the `CheckSCTPModuleLoadedOnNodes` check here, because that's now redundant with the new "should allow creating Pods, Services, and NetworkPolicies with SCTP ports" test.

#### IV. Drop the SCTP-only periodic job(s?)

There are one or more periodic jobs testing the `[Feature:SCTP]` tests left over from when SCTP was alpha and not enabled by default (eg https://testgrid.k8s.io/sig-network-gce#ubuntu-gce-basic-sctp), but they can go away now.

**Comment by theunrealgeek** [comment]
@danwinship I'm new to contributing to kubernetes, so I might be slow and take some time to get this done, but would love to take it up. Seems pretty interesting

**Comment by cmluciano** [comment]
@FreeZhang61 @theunrealgeek Have either of you had some time to take a look at this issue? There are a few pieces that can probably be worked in parallel.

**Comment by theunrealgeek** [comment]
@cmluciano Since @FreeZhang61 assigned it to himself, I moved on :sweat_smile: However if it is still up for grabs, I can spend some time this coming weekend

**Comment by sorushsaghari** [comment]
as a super beginner can I work on it ??

**Comment by theunrealgeek** [comment]
I'm currently working on `II. Simplify the core [Feature:SCTP] tests` and also trying to get my local setup working to run the integration tests. Not sure if I should be using `kubetest` as mentioned in the [e2e tests doc](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/e2e-tests.md) or `kubetest2` which is linked from the [deprecation warning on the `kubetest` repo readme](https://github.com/kubernetes/test-infra/tree/master/kubetest#deprecation-notice).

 Also @danwinship for

> The three tests should be replaced with a single simpler test, something like "should allow creating Pods, Services, and NetworkPolicies with SCTP ports". (It probably doesn't belong in service.go any more at this point though.) It doesn't try to use the Pod/Service/NetworkPolicy it creates, it just confirms that (a) you can create them, and (b) the result of CheckSCTPModuleLoadedOnNodes doesn't change from the start of the test to the end of the test

Do we want the test to look like
- check sctp module status
- create Pod
- check sctp module status
- <same for service and network policy> 

Or do we want
- check sctp module status
- create Pod
- create Service
- create NetworkPolicy
- check sctp module status 
?

And the current tests create Pods with container port and host port, services with regular ports and ClusterIP, should this diversity be captured in the new test ?

**Comment by danwinship** [comment]
> Not sure if I should be using `kubetest` as mentioned in the [e2e tests doc](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/e2e-tests.md) or `kubetest2` which is linked from the [deprecation warning on the `kubetest` repo readme](https://github.com/kubernetes/test-infra/tree/master/kubetest#deprecation-notice).

Sounds like `kubetest` still works but is just not getting new features, so I'd say use that.



> Or do we want
> 
>     * check sctp module status
>     * create Pod
>     * create Service
>     * create NetworkPolicy
>     * check sctp module status

like that

> And the current tests create Pods with container port and host port, services with regular ports and ClusterIP, should this diversity be captured in the new test ?

It should mostly test the same objects the current tests do, just all consolidated together

**Comment by theunrealgeek** [comment]
@danwinship I have been trying to get the tests to run locally and I followed the [this readme](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/e2e-tests.md#building-kubernetes-and-running-the-tests) so basically running

```
kubetest --build
sudo PATH=$PATH hack/local-up-cluster.sh
```
On another terminal
```
export KUBECONFIG=/var/run/kubernetes/admin.kubeconfig
kubetest --test --test_args="--ginkgo.focus=\[Feature:SCTP\]" --provider=local
```

And I see the following error as the tests are being run (the error is similar for all the existing tests for `Feature:SCTP`)
```
• Failure [0.115 seconds]
[sig-network] SCTP [Feature:SCTP] [LinuxOnly]
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/network/framework.go:23
  should create a Pod with SCTP HostPort [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/network/service.go:3801

  Dec 16 22:37:30.561: Unexpected error:
      <*errors.StatusError | 0xc0012721e0>: {                                                                                                                                                  
          ErrStatus: {                                                                                                                                                                         
              TypeMeta: {Kind: "", APIVersion: ""},                                                                                                                                            
              ListMeta: {                                                                                                                                                                      
                  SelfLink: "",                                                                                                                                                                
                  ResourceVersion: "",                                                                                                                                                         
                  Continue: "",                                                                                                                                                                
                  RemainingItemCount: nil,                                                                                                                                                     
              },                                                                                                                                                                               
              Status: "Failure",                                                                                                                                                               
              Message: "Pod \"hostexec-127.0.0.1-6bk29\" is invalid: spec.containers[0].securityContext.privileged: Forbidden: disallowed by cluster policy",                                  
              Reason: "Invalid",                                                                                                                                                               
              Details: {                                                                                                                                                                       
                  Name: "hostexec-127.0.0.1-6bk29",                                                                                                                                            
                  Group: "",                                                                                                                                                                   
                  Kind: "Pod",                                                                                                                                                                 
                  UID: "",                                                                                                                                                                     
                  Causes: [                                                                                                                                                                    
                      {                                                                                                                                                                        
                          Type: "FieldValueForbidden",                                                                                                                                         
                          Message: "Forbidden: disallowed by cluster policy",                                                                                                                  
                          Field: "spec.containers[0].securityContext.privileged",                                                                                                              
                      },                                                                                                                                                                       
                  ],                                                                                                                                                                           
                  RetryAfterSeconds: 0,                                                                                                                                                        
              },                                                                                                                                                                               
              Code: 422,                                                                                                                                                                       
          },                                                                                                                                                                                   
      }                                                                                                                                                                                        
      Pod "hostexec-127.0.0.1-6bk29" is invalid: spec.containers[0].securityContext.privileged: Forbidden: disallowed by cluster policy                                                        
  occurred
```

Is there some config I need to override for the default cluster policy to allow running of these tests ?

**Comment by danwinship** [comment]
hm... looks like maybe you need to pass an extra variable to local-up-cluster?

    sudo PATH=$PATH ALLOW_PRIVILEGED=true hack/local-up-cluster.sh

**Comment by theunrealgeek** [comment]
Just a quick update, I haven't been able to spend time on this last couple days, however should have something out in a PR soon.

**Comment by jayunit100** [comment]
I'm glad we suggested moving out of service.go.... 

Ok if we break out the network policy tests into a separate issue that can be fixed in the netpol suite ? 

Or do we want network policy tests to specifically live in an SCTP test suite like test/e2e/networking/sctp_integration_test.go ?

**Comment by theunrealgeek** [comment]
Since no one else has taken this up yet, I am still interested in getting this to completion once I get some time to spend on it

**Comment by danwinship** [comment]
> Ok if we break out the network policy tests into a separate issue that can be fixed in the netpol suite ?

I mean, someone could do a PR that fixed just that part of this issue, but I'd still like to be tracking the complete situation here

**Comment by theunrealgeek** [comment]
I should start getting some free time starting June and get PRs towards fixing this issue

**Comment by theunrealgeek** [comment]
@danwinship Hey (after a long time). So I have a WIP diff that looks like https://github.com/kubernetes/kubernetes/compare/master...theunrealgeek:sctp_tests_task_2_96717?expand=1 which is basically addressing 2 by creating a new test and combines the individual pieces from the existing tests. 

I'm having some trouble with testing this locally. I'm using a sequence of commands like

```
kubetest --build
sudo PATH=$PATH ALLOW_PRIVILEGED=true ./hack/local-up-cluster.sh

...

export KUBECONFIG=/var/run/kubernetes/admin.kubeconfig
kubetest --test --test_args="--ginkgo.focus=\[Feature:SCTP\]" --provider=local
```

and the first part of the test with a regular service+pod seems to work, but when the test is creating a pod with hostport, the pod is unable to come up with the following error

```

Events:    
  Type     Reason                    Age                From               Message    
    ----     ------                    ----               ----               -------    
      Normal   Scheduled                 24s                default-scheduler  Successfully assigned sctp-6825/hostport to 127.0.0.1        
        Warning  FailedCreatePodSandBox    24s                kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to start sandbox container for pod "hostport": Error response from daemon: driver failed programming external connectivity on endpoint k8s_POD_hostport_sctp-6825_d79b16bc-84cd-4568-ba78-85d4a71f05c5_0 (baa99ecc59bc43fbb0791dcdc33fbcb812e1e4bd3dca1fe429cfb6ed97f9c9bb): Error starting userland proxy: protocol not supported                        
          Warning  FailedCreatePodContainer  10s (x2 over 24s)  kubelet            unable to ensure pod container exists: failed to create container for [kubepods besteffort podd79b16bc-84cd-4568-ba78-85d4a71f05c5] : error while starting unit "kubepods-besteffort-podd79b16bc_84cd_4568_ba78_85d4a71f05c5.slice" with properties [{Name:Description Value:"libcontainer container kubepods-besteffort-podd79b16bc_84cd_4568_ba78_85d4a71f05c5.slice"} {Name:Wants Value:["kubepods-besteffort.slice"]} {Name:MemoryAccounting Value:true} {Name:CPUAccounting Value:true} {Name:IOAccounting Value:true} {Name:TasksAccounting Value:true} {Name:DefaultDependencies Value:false}]: Unit kubepods-besteffort-podd79b16bc_84cd_4568_ba78_85d4a71f05c5.slice already exists.
```

Do you know how I can enable SCTP on kubeproxy for local testing?

P.S I'm running Fedora 34 with `systemd-resolvd` and I also had to hack `local-up-cluster.sh` to point kubelet's `resolv.conf` at `/run/systemd/resolve/resolv.conf` for the cluster to even come up.

**Comment by danwinship** [comment]
> Do you know how I can enable SCTP on kubeproxy for local testing?

I don't think kube-proxy is the problem; it sounds like your network plugin doesn't support SCTP.

I'm pretty sure `kind` supports it so you could try with that...

**Comment by aojea** [comment]
> > Do you know how I can enable SCTP on kubeproxy for local testing?
> 
> I don't think kube-proxy is the problem; it sounds like your network plugin doesn't support SCTP.
> 
> I'm pretty sure `kind` supports it so you could try with that...

check the sctp kernel modules is loaded or try to load if for the tests `modprobe -b sctp`

**Comment by theunrealgeek** [comment]
Cool, let me try with kind then

**Comment by mgutierrez98** [comment]
Hello is anyone currently working on this?

**Comment by theunrealgeek** [comment]
@mgutierrez98 I am currently working on this issue, have a rough PR out and still addressing feedback.

/remove-lifecycle stale

**Comment by RajivTS** [comment]
The linked PR appears to be closed, so is it still being worked on? If not then I would like to pick it up.
@danwinship @theunrealgeek

**Comment by theunrealgeek** [comment]
@RajivTS Please take it up and feel free to take anything you like from my PR as well.

**Comment by RajivTS** [comment]
/remove-lifecycle rotten
/assign

Thanks @theunrealgeek.

**Comment by MorrisLaw** [comment]
Hey @RajivTS @AkshayPS12 any progress on this PR or is this open for anyone to work on it?
